src/js/hooks/useNavigation.jsx

```diff
@@ -0,0 +1,107 @@
+const useNavigation = ({
+  newRowFocusColumns,
+  existingRowFocusColumns,
+  tableData,
+  setFocusId,
+  setFocusIndex,
+  addEmptyRow,
+  productCode,
+  cycleCountId,
+}) => {
+  const getNextFocus = (columnId, rowIndex) => {
+    const currentIndex = newRowFocusColumns.indexOf(columnId);
+    const remainingColumns = newRowFocusColumns.slice(currentIndex + 1);
+    const newRowIndex = tableData[rowIndex + 1];
+
+    if (!remainingColumns.some(col => existingRowFocusColumns.includes(col))) {
+      if (!newRowIndex) {
+        addEmptyRow(productCode, cycleCountId);
+        return { newColumnId: newRowFocusColumns[0], newRowIndex: rowIndex + 1 };
+      }
+
+      if (newRowIndex?.id.includes('newRow')) {
+        return { newColumnId: newRowFocusColumns[0], newRowIndex: rowIndex + 1 };
+      }
+      return { newColumnId: existingRowFocusColumns[0], newRowIndex: rowIndex + 1 };
+    }
+
+    if (tableData[rowIndex]?.id.includes('newRow')) {
+      return { newColumnId: newRowFocusColumns[currentIndex + 1], newRowIndex: rowIndex };
+    }
+    return {
+      newColumnId: existingRowFocusColumns[existingRowFocusColumns.indexOf(columnId) + 1],
+      newRowIndex: rowIndex,
+    };
+  };
+
+  const getPreviousFocus = (columnId, rowIndex) => {
+    const currentIndex = newRowFocusColumns.indexOf(columnId);
+    const previousColumns = newRowFocusColumns.slice(0, currentIndex).reverse();
+    const previousRow = tableData[rowIndex - 1];
+    let newColumnId = newRowFocusColumns[currentIndex - 1];
+    let newRowIndex = rowIndex;
```

Comment:
just my opinion, not a change request.
Can we avoid mutating data in that case?